### DISCUSSION TITLE
Depth respecting types on local API

### DISCUSSION THREAD

**Discussion by AndKenneth** [payloadcms/payload#8229] [Feature Requests & Ideas]
Hey there, I've been using the local api in 3.0, and I've run into a bit of a pain point when it comes to the depth of relationship fields.

Lets say I have a type like this

```ts
export interface Profile {
  id: number;
  name: string;
  title: string;
  profileImage?: (number | null) | Media;
  bio?: string | null;
  updatedAt: string;
  createdAt: string;
}
```

The type `profileImage?: (number | null) | Media` means that whenever I retrieve profile using payload.find(), the returned documents will almost always return something with the type of Media, but the return type of payload.find() does not know this and types the result as `number | Media | null | undefined`.

This means that for every relationship, I have to check if the relationship exists, and do some type guarding logic to see if it's actually the ID (number) or if it's an object. At that point I could either return null on the items that are ids, or do seperate fetching logic to merge it in manually.

Ideally, the payload.find() function should know what it's returning based on depth, and therefore be represented in it's return type.

I think using gql even though I have access to the local api is a workaround, but feels like adding quite a bit of complexity for something that could be done fully within the local API.

Let me know what you think and if such a thing could potentially be on the roadmap!

**Comment by HriBB** [comment]
This would hugely improve DX ;)

**Comment by r1tsuu** [comment]
Hey @AndKenneth @HriBB! This is actually planned and I added it to the roadmap

**Reply by AndKenneth** [reply]
Awesome thanks! Can't wait for this.

**Comment by r1tsuu** [comment]
Opened a PR - https://github.com/payloadcms/payload/pull/9782
Take a look

**Reply by V1RE** [reply]
Cant wait for this to be merged! Any blockers still? Would be happy to help out if possible!

**Reply by willemmulder** [reply]
Great stuff!

**Comment by AndKenneth** [comment]
I was having a think through this after seeing the PR. While depth is the way that payload works right now and I'll definitely use the PR once in a release, I'm wondering if an `include` would be a more general solution. I think this would also work much better with types rather than hardcoding integers as typescript seems to make you do. Instead of saying "I want all the relationships to this depth", I would say, "these are the specific relationships I want". This is the way that prisma does it, and generally has been a bit more comfortable to work with. Doesn't drizzle also work kind of like this?

**Reply by willemmulder** [reply]
Although they are separate issues (the PR is about what type of field is returned, depending on the depth setting in the query, and your comment is about how we could make depth more granular), I think that would make perfect sense to me and also make queries much more efficient. Now we have to choose between none (depth 0) or all (depth 1) which isn't great. Interestingly, the GraphQL API might already support including arbitrary subpaths. 

Could you maybe make a separate issue?

**Comment by jxh0414com** [comment]
any update on this? i would love to have this.

**Comment by gavinmorrow** [comment]
What have people been doing as workarounds while waiting for this?

**Reply by maxsynnott** [reply]
Not clean but I use depth: 0 almost everywhere and use this helper to get the correctly typed id:

```ts
type PayloadRelationship = string | { id: string };

export const extractPayloadRelationshipId = (
	relationship: PayloadRelationship,
) => (typeof relationship === "string" ? relationship : relationship.id);
```

Would love to see progress on this.

**Reply by jxh0414com** [reply]
There is a extractID from payload/shared (this is in payload package) which is basically doing what you’re doing

**Reply by maxsynnott** [reply]
Ooh thanks, never found that 👌

**Reply by MurzNN** [reply]
On my websites the default depth = 0, but in many queries I pass custom depth = 1, 2, and even 3 in some cases.

To workaround TypeScript errors and additional depth-related checks and conditions, I composed helper types for each depth like this in a file like `src/types/collectionsWithDepth.ts`:
```ts
import {
  Offer,
  User,
} from '@/payload-types'

export interface UserDepth0 extends User {
  profileImage: Exclude<User['profileImage'], User>
}

export interface UserDepth1 extends User {
  profileImage: Exclude<User['profileImage'], string>
}

export interface OfferDepth0 extends Offer {
  siteId: Exclude<Offer['siteId'], SiteContext>
  owner: Exclude<Offer['owner'], User>
  request: Exclude<Offer['request'], Request>
  organization: Exclude<Offer['organization'], Organization>
}

export interface OfferDepth1 extends Offer {
  owner: Exclude<Offer['owner'], string>
  request: Exclude<Offer['request'], string>
  organization: Exclude<Offer['organization'], string>
}

export interface OfferDepth2 extends Offer {
  owner: UserDepth1
  request: RequestDepth1
  organization: OrganizationDepth1
  conversation: ConversationDepth1
}
```

and manually cast the correct type on every query like this:
```ts
    let user = (await this.p.find({
      collection: 'users',
      where: {
        username: {
          equals: username,
        },
      },
      depth: 1,
    }).then((res) => res.docs[0])) as UserDepth1
```

It is a lot of burden to manage this, but I still can't find a better workaround for this :disappointed: 

If someone invented something better, please share!
